### ISSUE TITLE
Machine Learning Roadmap

### ISSUE THREAD

**Issue by lindahua** [JuliaStats/Roadmap.jl#11]
Currently, the development of machine learning tools are in several different packages without little coordination. Consequently, some efforts are repetitive, while some important aspects remain lacking. 

Hopefully, we may coordinate our efforts through this issue. Below, I try to outline a tentative roadmap:
- Generalized Linear Models
  - Linear Regression
  - Logistic Regression
  - Lasso, Elastic Net, and its variants 
  - Stochastic Gradient Descent
  
  Current efforts: [GLMNet](https://github.com/simonster/GLMNet.jl), [GLM](https://github.com/JuliaStats/GLM.jl), [Regression](https://github.com/lindahua/Regression.jl)
- Support Vector Machines
  
  Current efforts: [SVM](https://github.com/johnmyleswhite/SVM.jl), [LIBSVM](https://github.com/simonster/LIBSVM.jl)
- DimensionalityReduction
  - PCA
  - ICA
  - CCA
  - Linear Discriminant Analysis
  - Kernel-based methods
  
  Current efforts: [DimensionalityReduction](https://github.com/johnmyleswhite/DimensionalityReduction.jl)
- Non-negative Matrix Factorization
  
  This may be categorized into dimensionality reduction. However, NNMF in itself has a plethora of methodologies, and thus deserves a separate package.
- Classification
  
  There are many techniques for classification. It may be useful to have multiple packages respective techniques (e.g. GLM, SVM, kNN), and have a meta-package _Classification.jl_ to incorporate them all.
- Clustering
  
  Current efforts: [Clustering.jl](https://github.com/johnmyleswhite/Clustering.jl)
- Many machine learning applications also require some supporting functionalities, such as performance evaluation, data preprocessing, etc. These can all go into [MLBase](https://github.com/lindahua/MLBase.jl)
- Probabilistic Modeling (e.g. Bayesian Network, Markov Random Field, etc)
  
  This is a huge field in itself, and may be discussed separately.

cc: @johnmyleswhite @dmbates @simonster @ViralBShah

Edit

---

I created an [NMF.jl](https://github.com/lindahua/NMF.jl) package, which is dedicated to non-negative matrix factorization.

Also, a detailed plan for DimensionalityReduction is outlined [here](https://github.com/JuliaStats/DimensionalityReduction.jl/issues/8).

**Comment by johnmyleswhite** [comment]
I agree with all of this. I've got a lot of prototype SGD code already.

I like the idea of meta-packages. If we're going to have Classification.jl, maybe Regression.jl should be a similar meta-package?

**Comment by jiahao** [comment]
I'm not an expert in this area, but I've been interested for awhile and am willing to help.

**Comment by lindahua** [comment]
@johnmyleswhite: Will you please move Clustering, SVM, and DimensionalityReduction over to JuliaStats? These are very basic for machine learning. I recently get some time to work on those.

For regression, when there are several quite different techniques implemented, it will make sense to make a meta package.

**Comment by johnmyleswhite** [comment]
I transferred Clustering and SVM over. I'm going to announce that I'm moving DimensionalityReduction over, then we can go ahead and make the move tomorrow.

**Comment by lindahua** [comment]
Also, I think it is important to separate packages that provide core algorithms and those integrated with DataFrames. 

We may consider to provide tools such that they can be worked nicely with machine learning algorithms. However, I think core machine learning packages should not depend on DataFrames -- which are not used as frequently in machine learning.

**Comment by johnmyleswhite** [comment]
I agree completely. I would very strongly prefer that we implement integration with DataFrames in the following way throughout all packages:
- Packages should always define algorithms that operate on `Vector{Float64}` and `Matrix{Float64}`.
- DataFrames.jl exposes a set of tools via formulas that translate between `DataFrame` and `Matrix{Float64}`.

This makes it easy to work with pure numerical data without any dependencies on DataFrames, while making it easy for people working with DataFrames to take advantage of the core ML algorithms by efficiently translating DataFrames into matrices.

**Comment by johnmyleswhite** [comment]
The only hiccup with what I just described is deciding whether the interfaces that mix DataFrames + ML should live. Arguably there should be one big package that does all of this by wrapping the other ML packages with a DataFrames interface.

**Comment by lindahua** [comment]
@johnmyleswhite are there issues of providing these in DataFrames.jl ?

**Comment by johnmyleswhite** [comment]
Providing what?

**Comment by lindahua** [comment]
Sorry, I seemed to misread part of your comments. I agree with your suggestions.

**Comment by lindahua** [comment]
Just that I am not sure whether we really another meta-package to couple DataFrames and ML, if the tools provided in DataFrames are convenient enough.

**Comment by johnmyleswhite** [comment]
You're right: we could encourage users to explicitly call the DataFrame -> Matrix conversion routines. That would simplify things considerably.

**Comment by johnmyleswhite** [comment]
The two main difficulties with this approach:
- Getting the community to adopt this kind of strategy consistently.
- Dealing with packages that legitimately need additional information to do their work. In GLM, for example, the entire model estimation steps need nothing more than access to the design matrix. But presenting the results in a convenient way requires access to the information about the original coefficient labels.

**Comment by lindahua** [comment]
For GLM, my consideration is to have two packages:
1.  A package that provides the core algorithms that only work with numerical arrays.
2. A higher-level package that builds on top of the core package that provides more friendly interface. (This package may depend on DataFrames)

**Comment by lindahua** [comment]
So this is basically your idea of having a higher-level package that relies on core ML packages + DataFrames to provide useful tools for analyzing data frames.

**Comment by IainNZ** [comment]
On phone right now, but weren't there some CART/Random Forest packages if not in METADATA then just mentioned in mailing list?
One thing about those is that they can use factors quite well, so I imagine would be directly dependent on DataFrames as that is the package-of-choice for representing that kind of data. So when talking about best practices etc. it might be worth keeping in mind that some packages might really be most efficiently made on top of DatFrames instead of the Matrix{Float64} abstraction

**Comment by lindahua** [comment]
Decision trees, by their nature, can work on heterogeneous data (each observation may be composed of variables of different kinds). For such methods, implementation based on DataFrames makes sense.
I don't mind a decision tree package depending on DataFrames.jl

There do exist a large number of machine learning methods (e.g. PCA, SVM, LASSO, K-means, etc) that are designed to work with real vectors/matrices. Heterogeneous data need to be converted to numerical arrays before such methods can apply. Packages that provide such methodologies are encouraged to be independent of DataFrames.

**Comment by johnmyleswhite** [comment]
You're right: there's a DecisionTree package.

To me, working with factors is actually a really strong argument for pushing a representation of categorical data into an earlier layer of our infrastructure like StatsBase. But we're actively debating ways to do this in JuliaStats/DataArrays.jl/issues/73.

If we could avoid some of the issues @simonster raised in his issue, I think it would be a big help to move the representation of categorical data closer to Julia's Base.

Also worth keeping in mind that nominal data is often worked with using dummy variables, which do fit in the `Matrix{Float64}` abstraction. That's actually how GLM handles those kinds of variables.

If DecisionTree.jl needs DataFrames.jl, I fully agree with Dahua: that's not a problem. But if it only needs a simpler abstraction, pushing things towards that simpler abstraction seems desirable.

**Comment by simonster** [comment]
There are some cases where `Matrix{Float64}` is too specific an abstraction. I have experimented with fitting point process GLMs in Julia, where the design matrix is theoretically expressible as a `Matrix{Float64}`, but it would require a huge amount of memory (for my models, probably >100 GB). On the other hand, it is easy to express the design matrix as an `AbstractMatrix{Float64}` that efficiently implements `A_mul_B!` and `At_mul_B!`. I wrote code that does this and directly minimizes the negative log likelihood via L-BFGS using NLopt, which fits my model in a reasonable amount of time with reasonable memory requirements, but I'm not sure what to do with this code, since the GLM package is still about 3x faster with a `Matrix{Float64}` (for the benchmark included with the GLM package with the same convergence criterion, excluding the non-negligible time to construct the ModelFrame).

As far as the model fitting interface for DataFrames, it would be cool if we could get this to work on top of `StatisticalModel`. Packages could implement:

``` julia
fit(::Type{MyModelType}, X::AbstractMatrix, y::AbstractVector, args...)
```

and DataFrames could implement:

``` julia
function fit{T<:StatisticalModel}(::Type{T}, f::Formula, df::DataFrame, args...)
   mf = ModelFrame(f, df)
   DFStatisticalModel(mf, fit(T, ModelMatrix(mf).m, model_response(mf), args...)
end
```

or similar. `DFStatisticalModel` could provide a wrapper that maps between coefficients and their labels when calling `coef`, `predict`, etc. Of course, doing this right requires that we have a reasonable StatisticalModel interface (#4) so that we can make the relevant functionality accessible for DataFrames.

**Comment by jiahao** [comment]
> There are some cases where Matrix{Float64} is too specific an abstraction.

This sounds a lot like the discussion we had in JuliaLang/IterativeSolvers.jl#2 a little while ago.

**Comment by andreasnoack** [comment]
@simonster GLM can use a sparse model model matrix, but I think you'll have to define your own subtype of `LinPred`.

**Comment by ViralBShah** [comment]
It would be great if as part of the roadmap, we can also plan to put together some large datasets in place, so that the community can work on optimizing performance and designing APIs accordingly. Having RDatasets is so useful, and something that makes large public datasets easily available for people to work with will greatly help this effort.

**Comment by lindahua** [comment]
@ViralBShah Good point. Datasets are important. I think we already have a MNIST package, we can definitely have more.

Just that we need to be cautious about the licenses that come with the datasets.

**Comment by johnmyleswhite** [comment]
There are surprisingly few large data sets that are publicly available. I'd guess that the easiest way to generate "large" data is to do n-grams on something like the 20 Newsgroup data set. Classifying one of the newsgroup against all the others is a simple enough binary classification problem that we can scale out to arbitrarily high size (in terms of features) by working with 2-grams, 3-grams, etc. Other useful examples might be processing the old Audioscrobbler data (http://www-etud.iro.umontreal.ca/~bergstrj/audioscrobbler_data.html) or something similar.

**Comment by ViralBShah** [comment]
We also have [CommonCrawl.jl](https://github.com/tanmaykm/CommonCrawl.jl). The point about the datasets is not as much to distribute them as julia packages, but to have easy APIs to access them, load them, and work with them. Often, I find that the pain of figuring out all the plumbing is enough to discourage people, and making the plumbing easy could get a lot more people to contribute.

**Comment by ViralBShah** [comment]
Perhaps not too big, but there's also the Netflix and MovieLens datasets - which could be made easier to access.

**Comment by johnmyleswhite** [comment]
The Netflix data set is illegal to distribute.

**Comment by gibiansky** [comment]
Question from an outsider - is there anything along the lines of Theano (from Python) in the works for Julia? Development of many deep learning models (RNNs) is sped up dramatically by AD-style software like Theano, and would allow the integration of deep learning into Julia much more easily...

**Comment by johnmyleswhite** [comment]
There are several AD tools in the works. Check METADATA for a few. There are also some GPU code-gen tools as well, including OpenCL.jl. Eventually it should be possible to combine those two into something like Theano.

**Comment by ccsv** [comment]
I would like to see something for association rule learning and neural networks

Other things needed should are gridsearch for finding hyperparameters (if not already implemented), Naive Bayes (with the +1 smoothing cases),  and  Restricted Boltzmann machine

**Comment by IainNZ** [comment]
Just pinged @benhamner, saw he listed another machine learning package and I'm hopeful for collaboration on shared interfaces

**Comment by BigCrunsh** [comment]
My feeling is that most of the packages are better suited in a separated JuliaML group; that would support the consistency. What is the disadvantage of having an own group?

**Comment by IainNZ** [comment]
No disadvantage, although I think they will be tightly linked if common functionality is re-used as much as possible. But the number of incompatible ML packages is starting to worry me...

**Comment by BigCrunsh** [comment]
Agree, me too... would be a chance to unify things.

**Comment by lindahua** [comment]
Lately, dimensionality reduction (MultivariateStats.jl), clustering (Clustering.jl), and nonnegative matrix factorization (NMF.jl) have been in a usable shape. 

However, there is one big area which is still in a messy status, that is, Generalized Linear Models / Regression. There have been several packages alone this line, implementing more or less similar functionalities, but they do not work with each other. 

I will open a new thread to discuss how we may proceed to unify the efforts in this domain.

**Comment by rofinn** [comment]
@lindahua I'm not sure what the status is on this topic anymore, but I noticed this thread while comparing MLBase.jl and MachineLearning.jl as utility libraries for experimenting with [Boltzmann.jl](https://github.com/dfdx/Boltzmann.jl). I see @benhamner hasn't responded to the issue from 6 months ago, but would it make sense to integrate the two packages together?  I'm particularly interested in some kind of pipeline api like [this](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) in one of the base libraries.

**Comment by johnmyleswhite** [comment]
@Rory-Finnegan: Right now I feel like there's no one with the time to take control of this project and give it the direction it needs. If you're feeling like demoing something that out, that seems like a good idea to me.

**Comment by benhamner** [comment]
@Rory-Finnegan just saw this - had missed the original ping (my Github notifications are spammed by internal Kaggle repos).

At this point, MachineLearning.jl has been a small playground I've touched here and there on the side as time permitted.

As @johnmyleswhite said, "right now I feel like there's no one with the time to take control of this project and give it the direction it needs." Definitely applies to me as well (at least for the visibility I have over the next 3 weeks, and likely longer). I've not looked closely at MLBase.jl yet, but need to. If you want to step up, go for it!

**Comment by amueller** [comment]
I'd love to have a look, but my julia skills are a bit below par ;) I'll see if I can find the time.

**Comment by ViralBShah** [comment]
What can we do as part of a GSoC project this summer to make progress here. Can we pick a small set of things to target to go further from where we are. I suspect there are lots of potential contributors, but we need someone to build a bit more of a framework before others can jump in. Someone here mentoring a GSoC student could get quite a bit of work done.

Should we also think of wrapping existing R and python libraries, and get the APIs right to start with, and then piecemeal, replace the underlying implementations? This python ML document was trending on HN today:

https://docs.google.com/a/fourthlion.in/document/d/1YN6BVdReNAYc8B0fjQ84yzDflqmeEPj7S0Xc-9_26R0/preview?sle=true

**Comment by rofinn** [comment]
So to keep the ball rolling on this topic I've created another repo with a README summarizing what I'm thinking I'd like in this base library. https://github.com/Rory-Finnegan/Learn.jl  If you have time please take a look and post feedback.  I'll admit that I'm coming from a sklearn background.

**Comment by swgregg** [comment]
I'm new to Julia and starting to use some VLM and NN in my graduate research in mechanical engineering.  I saw the reference to GSoC and am eager to potentially contribute to an ML or statistical learning project this summer.  Anyone know if there will be a Julia project along these lines for GSoC?

**Comment by johnmyleswhite** [comment]
I suspect there's not going to be an ML GSoC project since there's no one who's got time to mentor a student.

**Comment by johnmyleswhite** [comment]
@Rory-Finnegan: I like your proposal a lot. I think the best thing you bring up is that we can use "interfaces" to solve the biggest blocker we came across earlier: the lack of a coherent hierarchy that we could place most models into.

**Comment by lindahua** [comment]
Sorry for being inactive for months. 

Since I assumed the job of being an assistant professor last September, I feel that my life completely changes. I am now leading a group of PhD students and I find it is difficult to spare time to write codes myself. When I talked to a faculty member at Univ. of Toronto, I was told that "once become a faculty, fun life of coding is over" -- this is true, but a bit sad.

Another problem is that with the thriving of the entire deep learning business, many classical machine learning methods, like many we are discussing here, are quickly becoming irrelevant for the ML community. We should seriously reconsider the way going forward. I believe that logistic regression, linear regression, SVM classifiers, etc should no longer be considered as standalone procedures, instead they should be treated as building blocks to construct more sophisticated systems. Therefore, the interactions between them should be taken seriously.

**Comment by swgregg** [comment]
I see that Julia is no longer listed in the accepted organizations on GSoC (or maybe it was never listed this year and I was looking at last year previously).  I may still try to get involved with GSoC, however, I'm interested in contributing to ML in Julia either way.  My background is not CS and I have primarily just coded in Matlab for engineering projects.  That said, I am working on school projects right now for an algorithms class and may try to implement some of the project work in Julia to get a feel for it.

In my masters research I plan to use ML techniques to help perform condition monitoring on either wind turbines or hydroelectric turbines (depends on funding).  I'm working with Matlab now as I learn the statistics behind SVM and NN, however, I would like to use open source software for my research.  I will be spending at least part of the summer developing my coding skills and Julia seems like a good fit with opportunities for contributing at a fundamental level while deepening my understanding of ML.

I realize I am an outsider and haven't had enough time to wet my feet in this community.  That said, I am on spring break now and plan to start using Julia for a dynamic programming project due in a few weeks.  If anyone out there has input on how I may be able to help this summer (preferably in ML/statistics and taking into account my skill set) please let me know.  I was a senior controls engineer with 12 years experience before I decided to come back to school... so I don't necessarily need mentoring, just a point (or kick or shove) in the right direction and honest feedback.

**Comment by ViralBShah** [comment]
We will most likely have other ways to have the equivalent of GSoC funding this summer. We are working on this, and while nothing is firm, we will announce once ready. 

Cc: @alanedelman

**Comment by ViralBShah** [comment]
@swgregg IMO, a practical way to start off with a github repo for the project you are working on, and file issues against relevant packages as you run into missing functionality, roadblocks, or design issues.

**Comment by swgregg** [comment]
Thank you for the advise.  I will do just that. 

Date: Sun, 8 Mar 2015 05:35:32 -0700

**Comment by rofinn** [comment]
@lindahua So I definitely agree that part of the goal should be to standardize the interactions between models, since more and more people are stacking/combining different techniques (like using a dimensionality reduction alg in front of a classifier). The simplest way of dealing with this seems to be a model container type like Pipelines, in which the container is responsible for validating the interactions between models. If you have a better idea please let me know.

I'm not sure I agree that classic machine learning techniques are becoming irrelevant or that they shouldn't be used as standalone procedures. Especially since not everyone who uses these ML techniques are necessarily part of the ML community. Sometimes folks might just want to use a simple regression or SVM for their problems. Either way, I don't see any reason we can't support both approaches. :)

I'm still pretty new to deep learning so I could be missing an important step, but for deep belief networks, deep neural networks, etc couldn't we also just express them as a Pipeline? Specifically, that you'd build a deep learning architecture by subtyping Pipeline adding in whatever model units you want and then applying a fine tuning algorithm like backprop. We could probably also organize it so that Pipeline is a LearningModel, allowing for recursive nesting of Pipelines (once again, this would be made easier and cleaner with some kind of traits or interfaces system). Unfortunately, I'm not sure how this approach would relate to existing deep learning packages like [Mocha.jl](https://github.com/pluskid/Mocha.jl). Thoughts? FYI, I'm currently looking at deep learning architectures for my thesis, so I'll probably want to integrate my code with this package anyways. 

If you're okay with that approach, I could probably find some time this week to flesh out the Pipeline stuff (or w/e we want to name the container type).

**Comment by lindahua** [comment]
@Rory-Finnegan Classical machine learning techniques are still very important and widely used in many situations. What I was saying is that the focus of machine learning application has shifted gradually towards systems that integrate multiple components, and deep models are a case in point.

Hence, a sustainable strategy going forward should be to develop components such that they can nicely interact with each other through carefully crafted interfaces. As far as I can see, _graphical models_ and _neural networks_ are two most popular frameworks that allow one to put together a number of different components in a way that is mathematically valid.

**Comment by amueller** [comment]
@Rory-Finnegan The combined models in neural nets are usually trained in their entirety using backpropagation. The scikit-learn pipeline does not make any assumptions about differentiability, and is not able to backpropagate gradients. So I would see that as the major difference.
I agree about your point about machine learning outside of deep learning. I am pretty sure it will not and should not should go away anytime soon.

**Comment by SimonAB** [comment]
@amueller as someone starting to apply ML in the biomedical field I must point out the growing importance of inference in supervised learning (say, ensemble methods with decision trees for deep sequencing)... so I would disagree with the statement that ML outside of deep learning will or should go away, unless deep learning preserves full visibility and interpretability of feature importances. My anecdotal evidence suggests there is much excitement and democratisation of ML approaches beyond the core data-science crowd. I see this as an important strategic opportunity for Julia.

And with that in mind, I fully agree with @lindahua that it is extremely desirable to build in the ability to integrate multiple ML components from the outset, if nothing else to 'future proof' Julia's ML ecosystem.

**Comment by amueller** [comment]
My statement was missing a "not" whoops. I totally agree with you.

**Comment by ViralBShah** [comment]
It does make sense to think through the APIs and composability. There are plenty of implementations in R and python that we can wrap to start with, eventually replacing them with Julia implementations. Has anyone used @svs14 's Orchestra, which seems to have wrappers for scikits-learn and caret?

https://github.com/svs14/Orchestra.jl

**Comment by svs14** [comment]
I have used Orchestra ;)

I have not touched this for 3 months along with everything open source though due to some legal ambiguities in my previous internship, will work on it again from next week - so it's not dead.

For my purposes, I'm not concerned whether a machine learner is developed in a specific language, as long as I can compare + compose it together in a larger system I'm happy. If I truly need performance then I can gracefully degrade (in terms of effort of creating something from scratch compared to using a library) to write the respective learner in Julia.

Speaking along these lines, if it's not already considered, I would argue that pre-processing transformers such as near zero variance filtering found in caret can be pretty handy if not equally important to the learner in use in terms of a composable API. It's also easier this way if you enjoy applying grid-search to the pipeline itself, including omission of pre-processors.

Hope this adds to the discussion.

**Comment by ViralBShah** [comment]
Would it make sense to separate out Orchestra's wrappers around Python and R libraries into separate packages? I suspect they may receive more attention and can become common components in other projects.

**Comment by rofinn** [comment]
+1 I completely missed Orchestra when I was scanning through existing packages :( On the same line of thinking would it also make sense to see if existing julia packages (like DecisionTrees.jl, SVM.jl, etc) could be updated to support Orchestra's API? @svs14 I'd be happy to help you work on this if you're interested.

**Comment by ViralBShah** [comment]
In fact it was @aviks who showed me this package a few days back.

**Comment by svs14** [comment]
As a heads up, extracting Orchestra's scikit-learn + caret wrappers to packages that cover the full spectrum of each may require a design overhaul - right now the wrappers only target classifiers. Also the solutions I used were very specific to constraints I had at the time, I suspect there are better ways to wrap both libraries now (especially caret, I went through PyCall.jl to rpy2 as there was no direct functioning route back then). Orchestra's wrappers may be beneficial as some inspiration, to a wrapper library built from the ground up.

Orchestra's API is very unstable and immediately suited for only a limited sub-domain of machine learning. For instance I'm currently investigating/developing on handling the spectrum of learning settings including semi-supervised, transfer and multi-task learning which will not be backwards-compatible. As such, it'd probably be best not to have other packages depend on it.

I think there are a number of well thought out machine learning APIs targeting different priorities as evidenced in the excellent discussions within the Julia community, and through Python's scikit-learn, R's caret, Java's weka, and Go's golearn (I'd love to have wrapped all of these as it gives me their learners for free lol!). As long as the API is unstable, it'd probably be best for the API designer to consider inversion of control and build wrappers for each package, instead of having each package developer responsible for adhering to it. IMO, this makes it a lot easier to change the API at will without buy-in from dependents, along with not placing any unstable standards' effort on package developers.

@Rory-Finnegan , would be great to work with you! I really like your ideas in Learn.jl and this discussion, except I don't have your email/contact-details, you can ping me at svs14.41svs@gmail.com if you want.

**Comment by IainNZ** [comment]
Looking at MathProgBase.jl's design might be of interest too - its essentially a general interface for constrained optimization problems, with 11 registered packages using it and other experimental packages also using it as a way to plug in to the infrastructure. Its composable too: i.e. I can make a pseudo-solver that takes inputs via the interface, and then solves a series of subproblems through the interface.

**Comment by datnamer** [comment]
+1 on consideration for inference aside from deep learning- Important for growing computational political/social  science field.

**Comment by wildart** [comment]
I do not think that there is a big problem in designing a comprehensive ML framework. It is mainly a lack of commitment from the Julia ML community, including myself, in pushing forward with an initial design. I believe ML framework would evolve along with the language, and whether it will look as `Learn.jl` or  `MathProgBase.jl` or `scikit-learn`, it is really circumstantial. If we are going to wait for interfaces or traits implementation in Julia which would certainly enforce particular standard on a ML packages, it will only postpone development of a general ML framework.
I believe that providing common thin type hierarchy, in a manner of `Learn.jl` or `StatsBase.jl`, is enough to start development of various libraries for particular implementation of ML algorithms (even multiple). After all, a correct implementation of an ML algorithm is a pure scientific endeavor.
And there should be some packages with engineering thought behind, like MathProgBase or Orchestra, which would have wrapping and data pipelining (including utilities and supporting functionality) for implemented ML algorithms without any particular preference. I value such packages more that any state-of-the-art learning algorithm, because they provide more benefits for a larger community.
Let's push for some initial draft of a common ML interface that everybody will start to adopt. I like `Learn.jl` interface as an umbrella interface. It is based on a well known separation of ML algorithms, which could be gradually extended in particular implementations of ML algorithms, and in their turn integrated into the umbrella interface if necessary.

**Comment by ViralBShah** [comment]
I don't know that this will necessarily help, but there is the possibility of a position at MIT to push forward on this, if someone were interested to take it up full time. There are also some funds at NumFocus for Julia development, and this would qualify - but that would be for a much shorter duration. Perhaps someone who is focussing on this exclusively can be an anchor around which everyone can contribute.

**Comment by ViralBShah** [comment]
I feel like JuliaOpt has taken this approach of nailing down the APIs and building a flexible composable infrastructure. Of course, we had @mlubin and @IainNZ who anchored that work and many others joined. We need the same here.

**Comment by SimonAB** [comment]
@ViralBShah I agree. A full time position on this at MIT seems ideal given the importance of the field... among many pluses, this should also ensure face-to-face interactions with Julia core developers when low level changes would benefit other computationally demanding fields (e.g. BioJulia @dcjones)

**Comment by Sisyphuss** [comment]
As a PhD student on statistics and machine learning, I'll keep a close eye on this issue, and am willing to contribute to it.

By the way, I think what we need is a _Grammar of Machine Learning_.

**Comment by rofinn** [comment]
I look forward to someone being able to work full time on this, since I can only spare a few hours a week. In the mean time, I'm working with @svs14 on `Orchestra.jl` and maybe merging the common structure into `Learn.jl`. After we have the ensemble stuff refactored and working, I'll talk to the `Mocha.jl` folks about how deep learning should work as they seem to have a pretty popular approach.

**Comment by datnamer** [comment]
Would this project include frequentist and bayesian inferential models as part of the hierarchy? Perhaps  @dmbates ,  @scidom  and maybe @Fonnesbeck can chime in.

**Comment by rofinn** [comment]
@datnamer I'm inclined to suggest that frequentist and bayesian inferential models might make more sense as part of the [StatsBase.jl](https://github.com/JuliaStats/StatsBase.jl) package. I may include some wrapping functionality so that arbitrary models could be used as well, so long as they support the approriate methods.

**Comment by papamarkou** [comment]
Hi @datname, sorry for the slow reply, have been abroad over the last two weeks. Not sure where we should get with this in the long run - there is already a placeholder in `PGM.jl`. I think for now it is better to let the inferential modelling frameworks mature at their own independent pace given that they are at an infancy level. We can discuss merging efforts in the future, on the basis of a broader and richer codebase.

**Comment by datnamer** [comment]
@Rory-Finnegan  and @scidom - Makes sense

**Comment by amueller** [comment]
If you want to design a common interface, I think it is important to define the scope. I probably makes sense to leave graphical models, structured prediction and probabilistic programming out of scope.
But there are many other cases apart from classification and regression.
I'm not sure it makes sense to try to define a very strict interface, and starting with the actual algorithms as @wildart proposed might be more fruitful.

Just some API cases to consider:
- dataframes vs matrices as inputs
- multi-label, multi-output and multi-task prediction
- algorithm evaluation in grid-search, interfaces for metrics (allow out-of bag scoring, allow ranking metrics for both classification and regression, ....)
- semisupervised learning (how do you identify unlabeled points, now does cross-validation handle these)
- missing value handling
- categorical variable handling (hard with matrices, easier with dataframes)
- recommendation system interfaces (the interface is quite different from classification / regression)
- online and active learning interfaces
- reinforcement learning (in scope?)
- building pipelines (what are the interfaces here? Can we subsample data?)
- Regularization path algorithms and their interaction with cross-validation and grid-searches
- Last but not least: pipelines for online learning

This is part of the laundry list of API choices as well as unsolved / punted issues in scikit-learn ;)

**Comment by lindahua** [comment]
I think @ViralBShah's idea is important. Technically, there can be many approaches to make this successful. The real problem is that we lack a person that can dedicate to this and drive the progress for long enough.

**Comment by svaksha** [comment]
I would like to contribute too but the sheer number of packages scares me. It would be really nice if there was an experienced mentor willing to lead and guide the effort.

**Comment by rofinn** [comment]
@amueller I agree that it would help to define the scope of the API, before starting and that it would help to work with particular implementations in mind. However, I don't think all of those points need to be addressed immediately.

**Comment by rofinn** [comment]
Also, this thread is getting a little long now so I've opened a [chat](https://gitter.im/Rory-Finnegan/Learn.jl?utm_source=share-link&utm_medium=link&utm_campaign=share-link) on `Learn.jl` for folks who are interested.

**Comment by amueller** [comment]
Yeah I agree that not all points need to be addressed immediately and maybe for the moment it is more important to actually get something going.

**Comment by RaviMohan** [comment]
Shouldn't Reinforcement Learning be part of the ML roadmap? (sorry if I'm missing something obvious  Julia newbie here). Do game playing algorithms (like those based on MCTS)  fit into "machine learning"?

 I know a research team applying RL to gameplaying who'd appreciate a solid Julia RL library that can deal with large datasets (with SMDPs ,MAXQ etc and some other bits and pieces like a distributed MCTS implementation)  ,and am thinking of building something for them (and learn Julia into the bargain), but if some one's already working on it, perhaps as part of this roadmap, then I can probably fork/contribute to that rather than start from scratch.

**Comment by rinuboney** [comment]
Hi all,

I'm interested in contributing to achieving this roadmap as part of JSoC 2015. I was wondering if anybody is willing to mentor me on this project. The deadline is June 1st. That's too soon and a quick response would be great. If anybody has the time and are willing to do the same then please contact me(rinuboney@gmail.com) asap. I know Julia and machine learning. I can do this.

**Comment by ViralBShah** [comment]
@rinuboney It would help to put together a concrete proposal. There is a lot of discussion here, and it would be great to take a chunk of this as a JSOC project. If you can put something together based on the discussion here - what you will work on in the next 3 months, it will be easier to find a mentor.

**Comment by rinuboney** [comment]
I'm working on the proposal. It is accessible here: https://docs.google.com/document/d/1UBhEOqU1MMsjxfItDrg_XRViFgkZVSSvYACis6hT37o/pub

**Comment by ViralBShah** [comment]
Ok. Please mail it to juliasoc@googlegroups.com when it is ready.

**Comment by jiahao** [comment]
@rinuboney Thanks for your interest.

Your proposal is currently quite vague. It would be better if you could identify specific examples of ML techniques that you would want to run and show that the code is duplicated or redundant or has a complicated dependency stack. Let's say I'm interested in random forests - are there multiple current implementations? are the implementations too hard to use? or maybe not general purpose?

**Comment by rinuboney** [comment]
@jiahao Thanks for the feedback. 

There are multiple implementations for various ML models. Eg : RandomForests - https://github.com/bicycle1885/RandomForests.jl, https://github.com/bensadeghi/DecisionTree.jl, GLM - GLMNet, GLM, Regression etc. I believe that it's a good thing to have multiple implementations but the problem arises when one wants to try out different models. Machine learning is about experimentation. When faced with a classification or regression or clustering problem, the user should try out different models and use the model that gives the best performance. When the models are implemented in separate packages, they operate on different data types and have different API. So the code has to be rewritten to try out each model. This is where a base library can help by providing interoperability between different implementations through an API. With an API, the user can switch between models and algorithms instantaneously. The same problems are present when a user wishes to stack models from different packages. 

My proposal is more of coordination between the different machine learning packages scattered around so that they are can be used easily in a scikit-learn fashion. I thought that this is what the whole discussion was about.

I understand I should improve my proposal and I'm on it :+1:

**Comment by IainNZ** [comment]
@rinuboney make sure you look at https://github.com/Rory-Finnegan/Learn.jl

**Comment by rinuboney** [comment]
@IainNZ yeah my proposal is directly based on Learn.jl and this discussion.

**Comment by RaviMohan** [comment]
My (somewhat delayed) 2 cents. Rather than "unifying interfaces" etc, I'd rather have implementation of non existing functionality, which (imo) adds more value.  A comprehensive test suite for package X would provide much better value imo.

The ML ecosystem is quite young in Julia, and providing an API to rapidly evolving libraries is a bit premature  .In any case, the above list is hardly a real 'roadmap' ,it is just a (very comprehensive) list of ML topics - A 'roadmap' needs to have a sequencing of tasks and at least rough dates of completion to be meaningful. 

What we need at this stage (again just my 2 cents feel free to ignore) are solid, tested, scalable  _libraries_ with compelling usecases that will get more people actively using Julia on real world ML projects. Interfaces can be extracted when _mature_ libraries are aplenty  and are really not very valuable till then The arguments for interfaces/coordination etc aren't very convincing (to me, YMMV)

All that said, if one really really  really wants to work on interfaces, the first thing to do would be to build  a compelling _concrete_ case . Write an interface to a _specific_ (and limited) set of libraries X,Y,Z... so we can do _specific_ tasks A, B, C with 2 lines of code vs 20 (or whatever).

**Comment by rinuboney** [comment]
@RaviMohan Thanks for your feedback. I'm thinking more along this line. Consider an API is designed for the Julia ML packages. Then the various existing implementations in different packages could be unified. Existing libraries like scikit-learn, weka etc could also be wrapped in the API. Then the whole set of packages supporting the API can conceptually be used similar to a machine library like scikit-learn. I'll try to list some advantages to this approach:
- The packages supporting the API can be maintained separately.
- It takes less time than developing a single sold, tested and scalable ML library from scratch. 
- It provides a higher level of flexibility. Somebody can independently implement a new algorithm and comply with the API. Automatically, the new implementation is part of the Julia ML ecosystem which means it can be tested easily and added to existing projects without any hassle.
- mix and match implementations in different packages.

Then once the API is designed, I think the community should focus on solid, tested and scalable implementations in Julia. I believe this part can be done faster in a decentralized manner unified by the API as opposed to a centralized single library approach. If the community takes the API road, then in the end, the community will have a plethora of packages accessible through a unified API. If it's the solid, tested and scalable library road then community will have a good ML library just like all the other languages. I don't believe in a single perfect library for any purpose. trade-offs have to be made in all cases and a unified API and separate packages allows you to change the trade-offs immediately.

**Comment by RaviMohan** [comment]
This whole idea of designing APIs  first and then the implementations/wrappers rarely works in practice _in ML_ (I'm cynical from decades of experience in navigating APIs designed before practical experience from actual library implementation)

(imo) people who don't do the implementation don't (generally) design good APIs and the above argument is too theoretical and lacking a real world perspective. Nobody, in any ecosystem has ever come up with _one_ API that could wrap totally different libraries like scikit and  weka and have it be useful in the real world. If you pull it off you'll be the pioneer.

That said, don't let me discourage you. If you think you can design an API for wildly different packages and/or packages not yet made, go for it. I'm skeptical about success,but don't let that affect your enthusiasm. Follow your vision.

**Comment by rinuboney** [comment]
Although the implementations are wildly different, they have the same functions and can be used with the same API. Not exactly the same but, In Clojure(programming language) there is a library called [core.matrix](https://github.com/mikera/core.matrix). It is an API for working with matrices. Different implementations in native code,Java and Clojure support the API. Switching the implementation is trivial. I think it's possible to do something similar for ML packages. I know it's not possible to wrap up completely different libraries but a good number of them can be. Eg: scikit-learn and Go Learn have a similar API.

**Comment by RaviMohan** [comment]
"Not exactly the same"? core.matrix has _nothing_ to do with a generic ML API. Manylanguages have interfaces of some kind or the other.

as you well know core.matrix is a completely different beast from an ML API. You _can_ design wrappers for datastructures, with different level of abstraction and tradeoffs. We know this from the 70s!! By this logic every Java interface in existence is evidence that  a generic ML API can be designed?

If you have a _real world_ example of common APIs  for massively different ML libraries (scikit and weka as per the OP) , I'm all ears. Else this is a case of "should work in theory but never has  in practice" (imo). 

 I had a very talented friend try very hard for years to write a generic wrapper _just for Reinforcement Learning libraries_ and in the end it was an excessively generic mess no one wanted to use. However that doesn't mean someone else might not succeed tomorrow. As I said above, if you think you can do it, go for it, and more power to you for trying.

To repeat,  if someone thinks wrappers can be built that unify real world ML packages and "switch implementations trivially" ,that's great. 

 I am _very_ skeptical about this actually working, having worked on real world ML projects for years, but that shouldn't affect anyone's enthusiasm for the idea. I just don't think it is a workable idea is all.  I'll be glad to be proved wrong. 

I hope you succeed. Cheers.

**Comment by rinuboney** [comment]
Well I'm a student and I'm still learning what's possible and not possible. I just happen to like the idea and I'm willing to work on it.

**Comment by RaviMohan** [comment]
Good for you. 

We do need people to attempt the "impossible". That is how the world moves forward.  You don't need anyone's approval to do what you want to do.

Go for it. Good Luck.

**Comment by datnamer** [comment]
Check out  caret and mlr, "two attempts to create a unified framework across all types of algorithms for the various steps of machine learning in R (pre-processing data, training, testing, hyper-parameter optimization, etc.)." Sounds similar to @rinuboney 's stated goals. 

https://github.com/topepo/caret
https://github.com/berndbischl/mlr

**Comment by amueller** [comment]
> I don't believe in a single perfect library for any purpose. trade-offs have to be made in all cases and a unified API and separate packages allows you to change the trade-offs immediately.

For scikit-learn, many of the trade-offs are in terms of API.

On the other hand, a unified API is what brings people to scikit-learn, even before we had as many [and as fast] algorithms as we have now.
One of the reasons people convert from R to Python is that scikit-learn provides a unified and simple API.

**Comment by rinuboney** [comment]
They are similar. Haven't noticed it. I'll look into them in detail. @datnamer Thank you for pointing them out.

**Comment by rinuboney** [comment]
@amueller I'm a scikit-learn user and I really like the API. It makes ML really simple for beginners. I hope to make Julia ML packages accessible through a similar API.

**Comment by IainNZ** [comment]
I'd be pretty damn happy with a Julia `caret`

**Comment by rinuboney** [comment]
It would be awesome if I get a chance to work on it as part of JSoC. Please do have a look at my [proposal](https://docs.google.com/document/d/1UBhEOqU1MMsjxfItDrg_XRViFgkZVSSvYACis6hT37o/pub). Any feedback would help me refine the ideas.

**Comment by johnmyleswhite** [comment]
FWIW, I agree with @RaviMohan:

> My (somewhat delayed) 2 cents. Rather than "unifying interfaces" etc, I'd rather have implementation of non existing functionality, which (imo) adds more value. A comprehensive test suite for package X would provide much better value imo.

To give one simple example: building a reliable and efficient implementation of GBDT would easily take an entire summer.

We don't even have a unified interface for `AbstractArray` yet. In Python, you can run k-means on sparse matrices. In Julia, you can't. We're **really** far from being ready to provide unified ML libraries.

**Comment by rofinn** [comment]
@rinuboney Given the discussion that planning an API might be premature at this point and that your current proposal may be too vague, it might make more sense to identify something you would like to do in Julia, but are limited by the lack of abstraction between package. A simple example might be something like using an RBM from the Boltzmann.jl package as a feature extractor with an SVM from SVM.jl for classification (I know that is something that is pretty easy to do in packages like sklearn). Once, you have something working then maybe you could refactor the glue code into a common package like Learn.jl, with expectation that we'll need to iterate on it as the ML package ecosystem evolves.

**Comment by jiahao** [comment]
@rinuboney @Rory-Finnegan's advice is sound about finding specifics, as is @johnmyleswhite's, specifically regarding scope.

I reread your proposal and it is still vague. It still doesn't explain how bad the current situation is. Walk me through _one specific example_ of a ML technique that demonstrates why it is hard to do with Julia's packages today, and what you plan to do about it. Mock up some "after redesign" code showing how you would like to have the user code written.

**Comment by Evizero** [comment]
So what's the status on the machine learning community here. Have people moved on to other things and jumped ship?

**Comment by johnmyleswhite** [comment]
Many people are busy doing their day-to-day work with Julia rather than working on Julia.

**Comment by ViralBShah** [comment]
For many users of machine learning, I find that they are happily using one of the existing packages, or calling python, or have simply written the algorithms themselves.

**Comment by Evizero** [comment]
Allright then. I'll go ahead and do the things I need exactly how I need them. I think I have a good understanding of whats out there currently. It's probably the most fruitful approach anyhow

**Comment by lindahua** [comment]
I created this issue originally, with an aim to promote the efforts of building a cooperative ecosystem of machine learning packages. Over the past one and a half year, I found that the key problem is actually not lack of a unified interface, but lack of a dedicated force to drive the progress and lacks of consensus on certain basic issues.

Many of the contributors in this community are working on Julia or its packages on a voluntary basis, and they have their own commitments (_e.g._ projects that they are paid to work on). On the other hand, many of the packages (except for some fundamental ones, such as `StatsBase`, `Distributions`, and `DataFrames`) are only maintained by one or two contributors. As a consequence, it is very difficult to maintain steady progress over a long period. If we have one or two developers whose regular work is to develop & maintain these packages, the situation can be substantially improved. I encourage some of my students to use Julia on research projects where Julia is the best fit, and I expect them to become active maintainer of the packages that are directly pertinent to their research. However, this takes time. Also, we may resort to NumFOCUS for sponsorship, but not sure how this may work out.

Another significant issue that I had faced in many discussions is that people from different background tend to have notably different perspectives on how things should work. For example, people who work with data of moderate size (_e.g._ a table loaded from a typical CSV file, etc) in an interactive way would place much of the emphasis on friendly interactive API, while people who come from the background of designing machine learning algorithms (_e.g._ classifying or clustering a large set of vector samples) would consider performance as a more important factor -- that's partly the reason they move from R/Python/MATLAB to Julia. These people usually launch long-running experiments instead of working with the data in an interactive manner. Also, different communities seem to have different conventions on certain operations (_e.g._ whether to treat observations as columns or as rows). 

All these issues post challenges as we work towards a great ecosystem for machine learning. 

Considering the diversity of our community and that we are still in a very early stage, I think we should encourage diversity in our efforts. For certain well-established packages, we may encourage people to build their work on top of them, while for others, I think it doesn't hurt to have packages to approach the same problem/topic in different way. That being said, I think it remains advisable to coordinate the design of API between several closely-related packages -- just that we are still not at the position of designing a universally applicable interface.

**Comment by Evizero** [comment]
I figured as much. Well, I'm about 100 hours into Julia and I am hooked. Maybe this is a good time to introduce my self and state my intentions. I have already talked to my PhD supervisor and convinced him that Julia would be a good time investment for me and the group that I am in, and that I want to spend as much time on it as I can. Concretely that means that I will continue to work towards a ML Research Framework/Ecosystem with the help of a masters student and two BSc students. So I am here to help. I don't require funding or anything. I do have my own biased ideas how to do things though, but I think I am very much open to criticism. So I am very much interested in discussions and feedback.

A big interest of us are going to be evolutionary algorithms. So scalability matters to us very much and was the reason I even got here in the first place. That being said, the first thing I will tackle is some housekeeping. I do see your point that about the unified API, but I don't think it's that crippling. I have started to work on [SupervisedLearning.jl](https://github.com/Evizero/SupervisedLearning.jl), where the first thing I want to do is a simple front end for the most common supervised ML algorithms using established Libraries such as Regression.jl, Optim.jl, ForwardDiff.jl, OnlineAI.jl, and Mocha.jl as backend. Also I want to have some data and data-encoding abstractions that I will include in there. Any feedback is of course very welcome. Since SVM.jl is unmaintained I was thinking of taking over maintenance but after looking at the code I decided to do it from scratch in line with the SupervisedLearning interface.

So to sum up, first I will write the things needed to nicely compare and plot common things (Thats why I wrote [UnicodePlots.jl](https://github.com/Evizero/UnicodePlots.jl) ). The MSc Student is going to work on Feature selection which will make use of that, and one BSc will start to work on evolutionary algorithm in the feature selection context

**Comment by pluskid** [comment]
I'm copy & pasting a proposal about umbrella organization for deep learning here in case there would be people interested in.

> Sorry if this should really go to the julia-users mailing list. But it is easier to at authors of existing packages here on github. I want to initiate the discussion of the possibility of creating a JuliaDL umbrella organization, if enough people are interested in.
> 
> Currently we have several packages in Julia dedicated to deep learning (please comment if I missed any):
> - [Mocha.jl](https://github.com/pluskid/Mocha.jl): main contributors @pluskid, @benmoran, @stokasto, @nstiurca, @droidicus, ...
> - [KUnet.jl](https://github.com/denizyuret/KUnet.jl) and [Knet.jl](https://github.com/denizyuret/Knet.jl): @denizyuret
> - [Strada.jl](https://github.com/pcmoritz/Strada.jl): @pcmoritz
> - [Boltzmann.jl](https://github.com/dfdx/Boltzmann.jl): @dfdx
> 
> I think given the number of stars Mocha.jl received on github (3rd most star-ed Julia package, following IJulia and Gadfly.jl), there is no doubt that a fair amount of Julia users and developers have some interest in deep learning, not to mention that deep learning is becoming extremely important in machine learning and data analysis outside the Julia world.
> 
> So I'm expecting there will be further exciting development in deep learning related stuff in Julia, and I think it might be useful to have an umbrella organization. I think the goal of such an organization will be several folds:
> - Keep the mature existing packages within the umbrella, or at least have a centralized indexing into existing packages. I have to say none of the names in the packages mentioned above are very informative as their relation to deep neural networks.
> - Have a centralized place to discuss general plans and roadmaps for deep learning in Julia. Maybe also encourage collaboration and knowledge sharing, for a better future solution of deep learning in Julia. And when people want to jump in and contribute they will have an idea of the current status and identify interesting future projects.
> - It could also possibly raise more attention when we want to discuss about better collaboration with other parts of Julia. For example, deep learning is closely related to @JuliaStats (or maybe even should be a sub-category), as well as @JuliaGPU, and maybe @JuliaOpt.
> 
> Just for examples, I include several pointers here
> - Since almost all the non-trivial deep learning experiments need to run on GPUs, having a feature-complete Array type that lives on GPU memory is very important for both library developer and end users. Currently, at least Matlab, Python (e.g. via `pycuda`) and Lua (e.g. [cutorch](https://github.com/torch/cutorch)) have mature support for manipulating arrays on GPU directly. Currently we have [CUDArt.jl](https://github.com/JuliaGPU/CUDArt.jl), and related packages [CUBLAS.jl](https://github.com/JuliaGPU/CUBLAS.jl). But a lot of operations are still not supported. Julia already have built-in support for distributed array types. I think a good GPU Array type will be very nice to people even outside the deep learning world.
> - [Theano](http://deeplearning.net/software/theano/) is a very popular python tool for building deep learning libraries. It has automatic differentiation, so that you simply need to write your network in standard numpy expressions, and it will get compiled into computation code that could be executed on GPU very efficiently. It is very flexible and convenient to use, especially for researchers who needs to try a lot of non-standard operations and architectures. Quite a while (OK, 3 years) ago, there was a [discussion in the mailing list](https://groups.google.com/forum/#!msg/julia-dev/21JSSm8Z-gQ/IbrjGwRfZOYJ) about implementing a Theano equivalent in Julia. Julia seem to be an ideal platform to implement such stuff with its support for powerful macros and meta-programming, as well as LLVM backends. But yet it is not happening yet. I just want to keep this pointer here, so that when some people with relevant knowledge and interest could pick up this if they want to contribute. Just to mention, there is an alternative project to Theano, called [CGT](http://rll.berkeley.edu/cgt/), though still in Python.
> - Several month ago, [another very exciting thread](https://groups.google.com/forum/#!searchin/julia-dev/maleadt/julia-dev/doQUqQabC8A/mbDe8DSd2ykJ) was posted in the mailing list about some preliminary support of compiling Julia code directly to GPU devices. This means we could write GPU kernels (functions that runs on GPU) directly in Julia. For example, currently a large amount of code in Mocha.jl is still written in C++, because we need to write CUDA kernels in C++. I'm not blaming C++, I think it could be enjoyable experience once [Cxx.jl](https://github.com/Keno/Cxx.jl) is stable enough to run on Julia releases. But currently everything needs to go through a C API layer and it becomes painful when you want to support type generics for different floating point sizes, for example. Also, just to mention, you can already [compile python function to GPU](http://docs.continuum.io/numbapro/CUDAJit) with CUDAJit now.

**Comment by johnmyleswhite** [comment]
Will try to comment more thoroughly later, but I love the idea of an active community around deep learning in Julia.

**Comment by andreasnoack** [comment]
Great initiative. I think the most important requirement for an orginazation is that enough people commit to a common reposibility for the packages under the orginization. If that happens, then it is quite convenient with more committers to merge pull reuquests and help updating the packages e.g. when functions are deprecated.

A small suggestion: I'd consider naming the organization `JuliaDeepLearning`. Then there is no ambiguity about the focus.

**Comment by pluskid** [comment]
@andreasnoack Yes, maybe that is a better name. I agree with your comments on organizations. So maybe we should just put everything under JuliaStats if there is not enough people with interest / time commit to this. All we currently have are mostly some independent packages and some nice future ideas that nobody is working on.

**Comment by tkelman** [comment]
It sounds like there probably would be enough interest for a dedicated JuliaDeepLearning organization. It would have some requirements for interoperating with classical subcomponents that exist in JuliaStats. If there were a Julia equivalent of scikit-learn it would probably go in JuliaStats, but a julia equivalent of theano or cgt could go in a new JuliaDeepLearning org. At a bare minimum, start by moving Mocha there, and figure out the best next steps from there?

**Comment by Evizero** [comment]
I am assuming you - @pluskid - should have a pretty good picture of the state of the Julia deep learning community. So my guess is you probably have the most educated idea of what needs to be done to move it forward. We all know that deep learning is pretty much the most active subfield of ML right now, so I think it would be a good investment to make the Julia part more official.

Question is if there is enough of a community to maintain the packages if there is no explicit owner any more. MLBase is a good example for a package that I don't touch (even though it would make sense to add some code to it), simply because it takes a week to get a version tagging request replied to. Basically I don't think organizations are automatically a good idea; especially not if the author is actively maintaining his/her packages.

As a sidenote, I agree with @tbreloff and think a general JuliaLearn/JuliaML org would make more sense than moving the deep learning packages into JuliaStats; especially given the MLBase situation. I don't think the JuliaStats community has the resources at the moment to maintain ML packages to be frank. I don't want to step on anyone’s toes here. All the JuliaStats members that I had contact with were very nice and very helpful. I just think that they are busy with other things (such as Nullable Arrays) these days and don't have enough time to spend it on Machine Learning.

**Comment by lucasb-eyer** [comment]
Hi @pluskid, thanks for starting discussion. I'm currently working on a DL library in Julia which closely follows the design of Torch7, but makes use of Julia's features. It's not on github yet and progress is unfortunately slow because it's a side-project, my research is (still) in Theano. A friend of mine does a similar thing, so there definitely is interest. I also believe there's interest in the DL community at large, because Theano is suboptimal for RNNs and people generally don't like lua.

I agree that the current state of GPU array operations makes this task more painful than it ought to be, and a lot of work on this could probably be shared across DL packages.

PS: CGT looks promising, but it is not a _successor_ of Theano.

**Comment by tbreloff** [comment]
I have OnlineAI.jl, which extends OnlineStats.jl into neural nets and reservoir computing. I don't think it's appropriate for inclusion in a new organization, but there are pieces which overlap with other packages, and I think it would be great to have a unifying initiative for an MLBase that can support many different approaches to learning from data. 

My experience with many learning frameworks is that they tend to focus heavily on image classification and other similar (static) problems. I would really like to see something like the OnlineStats interface, which allows for both static (image classification, deep learning, etc) and dynamic (video analysis, time series, reinforcement learning, etc) modeling, allowing for analyzing both large distributed datasets and streaming data. Some of this exists already, and I hope we can create a best of breed base package to supply overlapping functionality.

**Comment by droidicus** [comment]
Just wanted to make sure you were aware of another SVM package in Julia called SALSA.jl: https://github.com/jumutc/SALSA.jl @jumutc

**Comment by Evizero** [comment]
> ... I think it would be great to have a unifying initiative for an MLBase that can support many different approaches to learning from data.

I agree. Actually I think MLBase is more of an "MLTools" in the sense that it provides design agnostic functionality. We should maybe think of collaborating on a common MLBase or MLAbstractions that does impose some design decisions such as function names. I know that I will sooner or later reach a point where I need to factor out a common base package for my stuff. I don't know much about OnlineStats.jl but I was thinking a little more high-level and really lightweight that evolves as we go along. Not everything falls under online learning, and probably not everything can be boxed in into the same kind of framework. Avoiding name collisions and settling on function names would be a good first step.

**Comment by dfdx** [comment]
@pluskid If you create organization, I'll be glad to join. Recently I added [cuRAND.jl](https://github.com/JuliaGPU/CURAND.jl) to JuliaGPU to support stochastic algorithms and currently am in a process of designing common library for unified CPU/GPU array programming - something similar to Theano/Torch7 (we should probably start a separate discussion about it). So if you are looking for people ready to contribute, include me to the list.

**Comment by denizyuret** [comment]
I'd be interested.  I am currently working on a Theano alternative model
compiler for Julia, it should take shape in the next couple of months.

best,
deniz

**Comment by tbreloff** [comment]
I think there is a lot of value in a consistent API, and I'm ready to put in some effort to make this roadmap a reality.  For the last few weeks I've been working on a very similar process with Plots.jl... putting a complex-but-lightweight interface into the plotting world.  I think the approach should be very similar for the ML community.

I propose that we create an organization JuliaLearn, and that we create a repo LearnBase.jl which will be home to both the design discussions and an implementation of what I describe (or something similar):  
- Design a bare minimum of verbs: `fit`, `fit!`, `predict`, `transform`, etc
- Design a method of mapping different data inputs to a consistent (and more verbose) "backend API".  i.e. `fit(model, dataframe)` would be mapped to a call like `fit(model, Any[convertColumn(c) for c in columns(dataframe)]; labels = names(df))`.  This way the user interface is simple, but we still retain full value in the data structure that was passed in, and final algorithms don't need special handling for DataFrames, etc
- Design traits or a type hierarchy which defines methods that a type of `LearningModel` must define... all probably require `fit`, online models require `fit!`, regression models require `predict`, etc.  Ideally I think the type hierarchy is not fully defined beforehand, but is implicitly defined given the common methods that a model implements.  With a robust API layer, we can use multiple dispatch to our advantage and forego many of the "type tree" problems that exist in other languages.
- Implement placeholder "linking code" that implements the "backend API", essentially converting calls from the backend API to existing packages.  For example, the user may call `fit(neuralnet, data...)` which gets converted into a call to the backend API `fit(neuralnet, processed_data)`, which then in turn will build a neural net in OnlineAI or Mocha or whatever default is chosen (likely chosen from installed packages with a priority list), and then return a wrapper around that Mocha object.  The user never knows about the differences in implementation between OnlineAI and Mocha, assuming they can both accomplish the request.
- Most backend packages are loaded as-needed, so that the REQUIRE file is small.  As such, LearnBase can be be used in many places without worrying about the massive dependency tree that plagues other projects.

This methodology has been (in my opinion) incredibly powerful for Plots.jl.  I have a simple, flexible API which can still access functionality from very different underlying packages, and requiring no cooperation from existing package authors.  It requires a little extra work up front to support a new backend package, but that is a much smaller effort than if that package would need to be re-written with a new interface, or to start a new package from scratch.  A user can make an API call which initially calls a python-wrapped library, but is then later replaced by a better julia implementation, with no change to their code.

There are two really important advantages to the approach that I described:
- The framework/API can be developed and designed without worry of package breakage or community turmoil.  Since you aren't requiring other packages to conform to any specific interface initially, it will be faster to achieve proof-of-concept and to iterate through design decisions. (contributors to LearnBase would not be dependent on other package authors for PR responses, etc)
- New packages can implement the "backend API" and instantly get the front end preprocessing for free, along with any other niceties that may be available (such as some variation of the `@stream` macro in OnlineStats.jl, or cross validation, or plotting recipes, etc).  This means the barrier to entry for brand new models/techniques is extremely low, and we may see more efficient development efforts in the future.

I am willing to take the lead on this effort, if you'll let me.  With a few :+1: I will form the org and get this started.

cc: @stefankarpinski @joshday

**Comment by simonster** [comment]
The first three points are already implemented. StatsBase defines [StatisticalModel and RegressionModel types](https://github.com/JuliaStats/StatsBase.jl/blob/master/src/statmodels.jl) along with methods for them. DataFrames [defines a fit method](https://github.com/JuliaStats/DataFrames.jl/blob/master/src/statsmodels/statsmodel.jl#L50-L56) that takes a subtype of StatisticalModel, a formula, and a DataFrame, converts the DataFrame into a design matrix according to the formula, and calls the `fit` method for the type with the model response and a design matrix.

I'm not sure there's a need for a separate organizations for statistics and machine learning. It may make sense to have a separate organization for deep learning, since it's substantially different from both stats and classical ML. But much of what is currently in JuliaStats could qualify as either statistics or ML. At least at this time I think it's better to do this work in a single organization.

**Comment by tbreloff** [comment]
To be clear, I see immense value in using as much of the current stats framework as is reasonable.  I think StatsBase would be one of the few required dependencies, and things like `StatsBase.fit` should be extended, not replaced.  There will be abstractions that are appropriate for online models, or deep neural nets, etc that are not appropriate for StatsBase, and those could fall into LearnBase (which is, at it's core, an extension of StatsBase)

As to whether LearnBase.jl should live in JuliaStats or JuliaLearn (or some other name the community agrees on), I can see both sides.  

Pros for JuliaStats:
- Already exists, people know about it
- Much overlap between stats and machine learning

Pros for JuliaLearn:
- Clean separation of owners (i.e. can have owners that don't impact JuliaStats)
- More focused community

I could be convinced either way...

**Comment by pluskid** [comment]
@Evizero Thank you for the comments! I'm starting to agree with you about the concerns of hosting projects under organizations. But I'm very glad to see that there are quite a few people with interest or already started working on theano / torch like system in Julia. We might consider creating an umbrella organization to host wiki pages pointing to those related projects and maybe host general discussions about deep learning libraries in Julia. May I ask, @lucasb-eyer, @dfdx, @denizyuret, when your new projects started to getting in shapes, could you come back and comment here? I think at that stage, we could consider creating such a repo. Having a wiki page summarizing different possible choices of libraries in Julia for deep learning will be at least very helpful for new users.

**Comment by Evizero** [comment]
> The first three points are already implemented

@simonster I don't think that is true. I have been playing with the idea of using `StatisticalModel` or `RegressionModel` as base class but they are simply not abstract enough. Not every learning model has coefficients and not every ML model that has coefficients has a probabilistic interpretation for them (i.e. things like `confint` don't always make sense). Here is the small difference between Machine Learning and Statistical Learning in my opinion.

@tbreloff I like the way you think, but I would really like to keep it much simpler and more realistic for now. I wouldn't got the Plots.jl route with the backends. For now we should just dictate the interface and class hierarchy, otherwise it is going to get ugly at one point or another. There should just be enough stuff in there that would be reasonable to expect new ML packages to follow. I think the two main goals should be
1. a user is able to import multiple ML packages without having name collisions occur
2. a user can expect similar things to have a similar interface that behaves similarly

I would also like to move my class-encoding code there (that builds on MLBase labelmap). Since it influences both our current efforts I'd suggest we just establish the package and get as many ML people in the loop as we can so that people can provide feedback. Since this package is a group effort it would make sense to me if it lived in an org. We can always move the package to JuliaStats later if it makes more sense, but for now let's just make some progress while we're motivated

EDIT: And to address the potential question of why not put this into MLBase: It doesn't even define the function name `accuracy` and the [PR that would add it](https://github.com/JuliaStats/MLBase.jl/pull/14) is sitting there unaddressed since April

**Comment by simonster** [comment]
If you don't define `confint` for a StatisticalModel nothing bad will happen. But there could be another level in the hierarchy if there is a perceived need.

I'm sympathetic to the concern than MLBase is not being sufficiently actively maintained, but it also looks like that PR failed its own tests.

**Comment by Evizero** [comment]
> If you don't define confint for a StatisticalModel nothing bad will happen. 

I get that, but that doesn't sound like a good solution

> But there could be another level in the hierarchy if there is a perceived need.

Yes, but I think this does need to be a group outcome. Since it is a problem that some people (which includes me) are currently actively concerned with I think it is a good time to brainstorm about this

> I'm sympathetic to the concern than MLBase is not being sufficiently actively maintained, but it also looks like that PR failed its own tests.

It's the not-even-replied-to part that bugs me, in the sense that anything non-trivial gets no reaction. I don't blame anyone who losses interest in contributing to Julia (or just a specific package) if no one even takes time to acknowledge the attempted contribution. I am not pointing fingers here. It's no one's fault. In fact, it's pretty cool that MLBase even exists to begin with. I think the StatsBase community is doing a tremendous job. But I do think it is a problem that needs to be addressed

I just think that given that a few people are currently very interested in actively working and improving Julia's ML aspects, that we should talk and address such problems that are crippling (for the lack of a better word) to the progress of the ML ecosystem.

But long story short, @tbreloff and I have started the discussion in LearnBase and we will try to code up a good solution. Anyone who is interested in the discussion or providing feedback is very welcomed

**Comment by johnmyleswhite** [comment]
FWIW, I think the best way to move forward is to punt on the abstraction layer problem for now (since we don't all agree on it and reaching group consensus is always extremely difficult) -- and instead focus on just nailing certain specific models. Simon's done amazing work to get regularized linear regression working well in pure Julia. It would be great to have similarly nice tools for things like kNN. I suspect it's easier to get people to collaborate (or at least offer useful feedback to one another) if everyone is coordinating on a single purely technical problem (e.g. how to make nearest neighbor search fast) that doesn't require people to come to consensus about purely aesthetic considerations.

**Comment by quinnj** [comment]
+1 John.
I haven't actively engaged in the thread here, but have followed since the beginning. I do think at this point, it's probably more productive for everyone to pick an area/model they're most interested in and really work on getting fast, feature-rich, usable ML code.
It might be more productive for everyone to plan on attending JuliaCon 2016 where we could plan a workshop(s) where ML Roadmap/Vision is discussed specifically. I think everyone meeting in person, coming with some solid code and ideas, would end up being much more productive in hashing out a coordinated vision for ML in Julia.

**Comment by Evizero** [comment]
Hmm maybe I have gone a little off track. I didn't know about the JuliaCon 2016 plans and I am very happy to hear about them (or at least the consideration)

But the two points I stated before still make sense to me

> -  a user is able to import multiple ML packages without having name collisions occur
> -  a user can expect similar things to have a similar interface that behaves similarly

I don't think settling on function names and defining them in a single place to avoid collisions is too far out there. I'm not talking about some fictive issues here. These are things that currently concern me in my efforts for SVMs. Some coordination, even if its just for exchanging ideas, is at least educating. I want to at least try and fail rather than not attempt at all.

**Comment by Evizero** [comment]
Let's leave it at this for now: It looks like @tbreloff and I will put our heads together and try to coordinate at least both of our current ML efforts in a meaningful way. Hopefully the outcome will be useful to others as well.

**Comment by ValdarT** [comment]
Hi

I am interested in what is the current state of ML ecosystem in Julia. By reading this (and other) issue(s) and having a look at the mentioned packages, it seems to me that:
- Hard work going on in JuliaML but Learn.jl will not be ready for use any time soon
- Orchestra.jl and SupervisedLearning.jl not maintained (I assume Learn.jl will fill their place in the future)
- ScikitLearn.jl maintained and works well but is not very actively improved/developed. (As it currently stands, it is more of an interface to the original code rather than a reimplementation in Julia.)

Are my impressions correct? If so, I assume people are not using Julia for day-to-day ML experiments like they use for example Python+scikit-learn? Or is there perhaps a ML ‘workbench’ package I missed?

**Comment by denizyuret** [comment]
Please check out https://github.com/denizyuret/Knet.jl

**Comment by rofinn** [comment]
@ValdarT I think most people using julia for "day-to-day ML" either use very specific packages for their use case (e.g. [Boltzmann.jl](https://github.com/dfdx/Boltzmann.jl), [BayesNets.jl](https://github.com/sisl/BayesNets.jl), [GaussianProcesses.jl](https://github.com/STOR-i/GaussianProcesses.jl), [Mocha.jl](https://github.com/pluskid/Mocha.jl)) or implement their own methods. I imagine the folks in the [JuliaML](https://github.com/JuliaML) organization are the most likely to come up with a good/cohesive julia framework all the different ML methods out there, but that's a pretty tough job.

**Comment by amueller** [comment]
Wow, JuliaML looks pretty great but also pretty ambitious. It has a much larger scope than scikit-learn and tensorflow combined... Is there any documentation on the "learn" package or a simple intro somewhere?

**Comment by ararslan** [comment]
Discussion of the JuliaML organization should take place in their roadmap: https://github.com/JuliaML/Roadmap.jl/issues. The focus of JuliaStats is more classical statistics, as the more ML-oriented packages in this organization are unmaintained (e.g. SVM and RegERMs).

**Comment by ViralBShah** [comment]
Locking and closing this issue so that discussion can continue in the right place: https://github.com/JuliaML/Roadmap.jl
